backport: random.h/.cpp cherry-picks up to #10821

xanimo opened the pull request:
This is rebased off of #3511, #3520, #3487. It includes commits to randcom.h/.cpp among others but is precursor to https://github.com/xanimo/dogecoin/tree/1.15.0-dev-14955-15224 and https://github.com/xanimo/dogecoin/tree/1.15.0-dev-14955-15224-15512-24926

xanimo commented:
Dropped the commits merged in #3511 and rebased the rest on 1.15.0-dev.

patricklodder commented:
I don't see any calls to `sync.h` code in the changes made by 6b08eca onward. What causes the dependency?

xanimo commented:
Interesting. Yeah those changes to sync.h/.cpp are extraneous for sure. Going to drop everything pre https://github.com/dogecoin/dogecoin/commit/6b08eca953de2c01732880aad266d00b1b063fad and include thread per your suggestion.

xanimo commented:
I modified https://github.com/dogecoin/dogecoin/pull/3521/commits/76e842061d6da29c9dbfec45a74905f6a8e92840 where mutex is introduced and included thread per your suggestion. I'm not sure if I should take ownership of that commit but I changed `cherry-picked from:` to `backported from:` in commit description.

patricklodder commented:
Looking much better now.

> I modified https://github.com/dogecoin/dogecoin/commit/76e842061d6da29c9dbfec45a74905f6a8e92840 where mutex is introduced and included thread instead

I think it needs to be included in another commit. I'll go through the remaining commits individually now that we have a mergeable proposal - we'll get it sorted.

xanimo commented:
I've separated that include thread and included prior to the introduction of this_thread::sleep_for in the test performance counter increments commit.

patricklodder reviewed (changes requested):
#### d5efa71

- cherry pick is missing include for `<chrono>` while introducing `std::chrono::high_resolution_clock::now()`- [original line missed](https://github.com/bitcoin/bitcoin/commit/f544094d5e1ee627ebd34542aead7bf1241fae19#diff-433f996b8149d6d6831ab1ee2978d4cc13dead5721c824175db806e4284ab15bR19)

#### 4ac1d0c1

- cherry pick is missing include of `<thread>` - [original line missed](https://github.com/bitcoin/bitcoin/commit/33f853d8d88917b145e2793bf8eb1b4e3790e7e7#diff-433f996b8149d6d6831ab1ee2978d4cc13dead5721c824175db806e4284ab15bR20) - doing the pick right also will make 038b0fc7 obsolete.

#### 75ea494f

- includes `validation.h` for no reason
- this is a bug fix on top of dba8b67 - probably be best to just squash this into that

#### 7a7f2b61

- is a bug fix on dba8b67
- is completely overwritten by 7fd793015, so this is a 100% pointless commit to backport. Recommend leaving it out.

#### 7fd79301 and 1c7be27e

- these are bug fixes on top of dba8b67 - probably be best to just squash these into that, we will never just revert these alone

#### e3b84fe1

- this is a horror of a commit because it fixes 30 issues all over the place - net, consensus, encoding, bench... so it expands the scope of this PR greatly
- Recommend just taking and doing the "inspired by" thingy: 
    - the definition of `size_t` in `compat.h`
    - changing the type of `NUM_OS_RANDOM_BYTES` to `int` in `random.h`.

patricklodder commented:
Additional note re: e3b84fe.

https://github.com/bitcoin/bitcoin/commit/f7dc99244c8e78dbd0196f612690efcc449c37dc does a really nice cleanup of the `ssize_t` definition for MSVC - maybe squash that in while we're at it?

xanimo commented:
I'm also looking at this atm. Not sure if I should squash these in https://github.com/dogecoin/dogecoin/commit/e3b84fe1ec99d4243907eab17fb0b203984c064e too?
https://github.com/bitcoin/bitcoin/pull/16158

patricklodder commented:
> Not sure if I should squash these in e3b84fe1ec99d4243907eab17fb0b203984c064e too?

Recommend putting that in its own PR, as it's beneficial to take that part and the additional 2 commits even without this PR ever going anywhere, even for an eventual 1.14.8. The documentation there is pretty good as well.

xanimo commented:
Force pushed to remove <stdint.h> and put the ssize_t definition in the right file src/compat/compat.h(accidentally grabbed their compat.h located in src/compat/compat.h

patricklodder commented:
Could you please remove the following items from the last commit so it's clean as far as this PR goes?

```
-include <array> and change buff from unsigned char[32] to static const std::array<unsigned char, 32> in bench/base58.cpp
-drop copy of buff to new variable b, begin() and end() and use data() and size() on buffer in bench/base58.cpp
-drop boost::assign::map_list_of for checkpointData in chainparams.cpp and change to std::list
-make bool operator a constant in net_processing.cpp IteratorComparator
-if msvc is defined include <Windows.h> and replace __asm with SecureZeroMemory() in support/cleanse.cpp
-change QUEUE_BATCH_SIZE to unsigned int in test/checkqueue_tests.cpp
```

i.e. only keep the changes to `compat.h` and `random.h` - everything else can be picked up separately.

patricklodder commented:
Awesome, I have one final request re: completeness. 

In #3522 you picked 1ec1602a, which is great, except #3522 depends on #3466 (BIP70 removal), which is huge and may take a longer time to get merged, whereas this commit will allow me to finish enabling c++17 compilation, and with that be a prerequisite to solve #3547 (protobuf > 22.x issues) even if we keep the payment requests for a while longer (hopefully not, but hey it's an uncertain world.)

Could you pick that as a final commit here? Then we have a complete interface to use, so that I can work on c++17 and we can start propagating the FRC as was done in bitcoin/bitcoin#14624 in parallel, while people make up their mind about BIP70.

xanimo commented:
Done :+1:

patricklodder reviewed (approved):
ACK - fully code reviewed and tested on arm64 macOS Sonoma (w/ clang 14) and x86_64 Ubuntu Jammy (w/ gcc 12)

Note: there are some indirect includes (e.g. for `<atomic>` in `random.cpp`). I think this should be fixed separately, as many of the files touched were not passing `iwyu` checks before these changes; fixing it properly would expand scope beyond the goal of this PR.
